File: swift-bridge/Sources/SpriteKitBridge/PhysicsBody.swift
import SpriteKit

@_cdecl("sk_physics_body_circle")
public func sk_physics_body_circle(_ radius: Double) -> UnsafeMutableRawPointer? {
    skRetain(SKPhysicsBody(circleOfRadius: CGFloat(radius)))
}

@_cdecl("sk_physics_body_circle_center")
public func sk_physics_body_circle_center(_ radius: Double, _ x: Double, _ y: Double) -> UnsafeMutableRawPointer? {
    skRetain(SKPhysicsBody(circleOfRadius: CGFloat(radius), center: CGPoint(x: x, y: y)))
}

@_cdecl("sk_physics_body_rect")
public func sk_physics_body_rect(_ width: Double, _ height: Double) -> UnsafeMutableRawPointer? {
    skRetain(SKPhysicsBody(rectangleOf: CGSize(width: width, height: height)))
}

@_cdecl("sk_physics_body_rect_center")
public func sk_physics_body_rect_center(_ width: Double, _ height: Double, _ x: Double, _ y: Double) -> UnsafeMutableRawPointer? {
    skRetain(SKPhysicsBody(rectangleOf: CGSize(width: width, height: height), center: CGPoint(x: x, y: y)))
}

@_cdecl("sk_physics_body_edge_loop_rect")
public func sk_physics_body_edge_loop_rect(
    _ x: Double, _ y: Double, _ width: Double, _ height: Double
) -> UnsafeMutableRawPointer? {
    skRetain(SKPhysicsBody(edgeLoopFrom: CGRect(x: x, y: y, width: width, height: height)))
}

@_cdecl("sk_physics_body_texture")
public func sk_physics_body_texture(
    _ textureHandle: UnsafeMutableRawPointer?,
    _ sizeW: Double, _ sizeH: Double
) -> UnsafeMutableRawPointer? {
    guard let texture: SKTexture = skBorrow(textureHandle) else { return nil }
    return skRetain(SKPhysicsBody(texture: texture, size: CGSize(width: sizeW, height: sizeH)))
}

@_cdecl("sk_physics_body_compound")
public func sk_physics_body_compound(_ rawBodies: UnsafeMutableRawPointer?, _ count: Int) -> UnsafeMutableRawPointer? {
    let bodies = skBodies(from: rawBodies, count: count)
    guard !bodies.isEmpty else { return nil }
    return skRetain(SKPhysicsBody(bodies: bodies))
}

@_cdecl("sk_physics_body_get_dynamic")
public func sk_physics_body_get_dynamic(_ bodyHandle: UnsafeMutableRawPointer?) -> Bool {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return false }
    return body.isDynamic
}

@_cdecl("sk_physics_body_set_dynamic")
public func sk_physics_body_set_dynamic(_ bodyHandle: UnsafeMutableRawPointer?, _ dynamic: Bool) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.isDynamic = dynamic
}

@_cdecl("sk_physics_body_get_allows_rotation")
public func sk_physics_body_get_allows_rotation(_ bodyHandle: UnsafeMutableRawPointer?) -> Bool {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return true }
    return body.allowsRotation
}

@_cdecl("sk_physics_body_set_allows_rotation")
public func sk_physics_body_set_allows_rotation(_ bodyHandle: UnsafeMutableRawPointer?, _ allows: Bool) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.allowsRotation = allows
}

@_cdecl("sk_physics_body_get_precise_collision")
public func sk_physics_body_get_precise_collision(_ bodyHandle: UnsafeMutableRawPointer?) -> Bool {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return false }
    return body.usesPreciseCollisionDetection
}

@_cdecl("sk_physics_body_set_precise_collision")
public func sk_physics_body_set_precise_collision(_ bodyHandle: UnsafeMutableRawPointer?, _ precise: Bool) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.usesPreciseCollisionDetection = precise
}

@_cdecl("sk_physics_body_get_pinned")
public func sk_physics_body_get_pinned(_ bodyHandle: UnsafeMutableRawPointer?) -> Bool {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return false }
    return body.pinned
}

@_cdecl("sk_physics_body_set_pinned")
public func sk_physics_body_set_pinned(_ bodyHandle: UnsafeMutableRawPointer?, _ pinned: Bool) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.pinned = pinned
}

@_cdecl("sk_physics_body_get_resting")
public func sk_physics_body_get_resting(_ bodyHandle: UnsafeMutableRawPointer?) -> Bool {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return false }
    return body.isResting
}

@_cdecl("sk_physics_body_set_resting")
public func sk_physics_body_set_resting(_ bodyHandle: UnsafeMutableRawPointer?, _ resting: Bool) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.isResting = resting
}

@_cdecl("sk_physics_body_get_friction")
public func sk_physics_body_get_friction(_ bodyHandle: UnsafeMutableRawPointer?) -> Double {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return 0 }
    return Double(body.friction)
}

@_cdecl("sk_physics_body_set_friction")
public func sk_physics_body_set_friction(_ bodyHandle: UnsafeMutableRawPointer?, _ friction: Double) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.friction = CGFloat(friction)
}

@_cdecl("sk_physics_body_get_charge")
public func sk_physics_body_get_charge(_ bodyHandle: UnsafeMutableRawPointer?) -> Double {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return 0 }
    return Double(body.charge)
}

@_cdecl("sk_physics_body_set_charge")
public func sk_physics_body_set_charge(_ bodyHandle: UnsafeMutableRawPointer?, _ charge: Double) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.charge = CGFloat(charge)
}

@_cdecl("sk_physics_body_get_restitution")
public func sk_physics_body_get_restitution(_ bodyHandle: UnsafeMutableRawPointer?) -> Double {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return 0.2 }
    return Double(body.restitution)
}

@_cdecl("sk_physics_body_set_restitution")
public func sk_physics_body_set_restitution(_ bodyHandle: UnsafeMutableRawPointer?, _ restitution: Double) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.restitution = CGFloat(restitution)
}

@_cdecl("sk_physics_body_get_linear_damping")
public func sk_physics_body_get_linear_damping(_ bodyHandle: UnsafeMutableRawPointer?) -> Double {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return 0.1 }
    return Double(body.linearDamping)
}

@_cdecl("sk_physics_body_set_linear_damping")
public func sk_physics_body_set_linear_damping(_ bodyHandle: UnsafeMutableRawPointer?, _ damping: Double) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.linearDamping = CGFloat(damping)
}

@_cdecl("sk_physics_body_get_angular_damping")
public func sk_physics_body_get_angular_damping(_ bodyHandle: UnsafeMutableRawPointer?) -> Double {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return 0.1 }
    return Double(body.angularDamping)
}

@_cdecl("sk_physics_body_set_angular_damping")
public func sk_physics_body_set_angular_damping(_ bodyHandle: UnsafeMutableRawPointer?, _ damping: Double) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.angularDamping = CGFloat(damping)
}

@_cdecl("sk_physics_body_get_density")
public func sk_physics_body_get_density(_ bodyHandle: UnsafeMutableRawPointer?) -> Double {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return 1 }
    return Double(body.density)
}

@_cdecl("sk_physics_body_set_density")
public func sk_physics_body_set_density(_ bodyHandle: UnsafeMutableRawPointer?, _ density: Double) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.density = CGFloat(density)
}

@_cdecl("sk_physics_body_get_mass")
public func sk_physics_body_get_mass(_ bodyHandle: UnsafeMutableRawPointer?) -> Double {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return 0 }
    return Double(body.mass)
}

@_cdecl("sk_physics_body_set_mass")
public func sk_physics_body_set_mass(_ bodyHandle: UnsafeMutableRawPointer?, _ mass: Double) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.mass = CGFloat(mass)
}

@_cdecl("sk_physics_body_get_area")
public func sk_physics_body_get_area(_ bodyHandle: UnsafeMutableRawPointer?) -> Double {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return 0 }
    return Double(body.area)
}

@_cdecl("sk_physics_body_get_affected_by_gravity")
public func sk_physics_body_get_affected_by_gravity(_ bodyHandle: UnsafeMutableRawPointer?) -> Bool {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return true }
    return body.affectedByGravity
}

@_cdecl("sk_physics_body_set_affected_by_gravity")
public func sk_physics_body_set_affected_by_gravity(_ bodyHandle: UnsafeMutableRawPointer?, _ affected: Bool) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.affectedByGravity = affected
}

@_cdecl("sk_physics_body_get_field_bitmask")
public func sk_physics_body_get_field_bitmask(_ bodyHandle: UnsafeMutableRawPointer?) -> UInt32 {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return 0 }
    return body.fieldBitMask
}

@_cdecl("sk_physics_body_set_field_bitmask")
public func sk_physics_body_set_field_bitmask(_ bodyHandle: UnsafeMutableRawPointer?, _ mask: UInt32) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.fieldBitMask = mask
}

@_cdecl("sk_physics_body_get_category_bitmask")
public func sk_physics_body_get_category_bitmask(_ bodyHandle: UnsafeMutableRawPointer?) -> UInt32 {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return 0 }
    return body.categoryBitMask
}

@_cdecl("sk_physics_body_set_category_bitmask")
public func sk_physics_body_set_category_bitmask(_ bodyHandle: UnsafeMutableRawPointer?, _ mask: UInt32) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.categoryBitMask = mask
}

@_cdecl("sk_physics_body_get_contact_test_bitmask")
public func sk_physics_body_get_contact_test_bitmask(_ bodyHandle: UnsafeMutableRawPointer?) -> UInt32 {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return 0 }
    return body.contactTestBitMask
}

@_cdecl("sk_physics_body_set_contact_test_bitmask")
public func sk_physics_body_set_contact_test_bitmask(_ bodyHandle: UnsafeMutableRawPointer?, _ mask: UInt32) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.contactTestBitMask = mask
}

@_cdecl("sk_physics_body_get_collision_bitmask")
public func sk_physics_body_get_collision_bitmask(_ bodyHandle: UnsafeMutableRawPointer?) -> UInt32 {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return 0 }
    return body.collisionBitMask
}

@_cdecl("sk_physics_body_set_collision_bitmask")
public func sk_physics_body_set_collision_bitmask(_ bodyHandle: UnsafeMutableRawPointer?, _ mask: UInt32) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.collisionBitMask = mask
}

@_cdecl("sk_physics_body_get_velocity_dx")
public func sk_physics_body_get_velocity_dx(_ bodyHandle: UnsafeMutableRawPointer?) -> Double {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return 0 }
    return Double(body.velocity.dx)
}

@_cdecl("sk_physics_body_get_velocity_dy")
public func sk_physics_body_get_velocity_dy(_ bodyHandle: UnsafeMutableRawPointer?) -> Double {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return 0 }
    return Double(body.velocity.dy)
}

@_cdecl("sk_physics_body_set_velocity")
public func sk_physics_body_set_velocity(_ bodyHandle: UnsafeMutableRawPointer?, _ dx: Double, _ dy: Double) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.velocity = CGVector(dx: dx, dy: dy)
}

@_cdecl("sk_physics_body_get_angular_velocity")
public func sk_physics_body_get_angular_velocity(_ bodyHandle: UnsafeMutableRawPointer?) -> Double {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return 0 }
    return Double(body.angularVelocity)
}

@_cdecl("sk_physics_body_set_angular_velocity")
public func sk_physics_body_set_angular_velocity(_ bodyHandle: UnsafeMutableRawPointer?, _ velocity: Double) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.angularVelocity = CGFloat(velocity)
}

@_cdecl("sk_physics_body_apply_force")
public func sk_physics_body_apply_force(_ bodyHandle: UnsafeMutableRawPointer?, _ dx: Double, _ dy: Double) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.applyForce(CGVector(dx: dx, dy: dy))
}

@_cdecl("sk_physics_body_apply_force_at_point")
public func sk_physics_body_apply_force_at_point(
    _ bodyHandle: UnsafeMutableRawPointer?,
    _ forceDx: Double, _ forceDy: Double,
    _ pointX: Double, _ pointY: Double
) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.applyForce(CGVector(dx: forceDx, dy: forceDy), at: CGPoint(x: pointX, y: pointY))
}

@_cdecl("sk_physics_body_apply_torque")
public func sk_physics_body_apply_torque(_ bodyHandle: UnsafeMutableRawPointer?, _ torque: Double) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.applyTorque(CGFloat(torque))
}

@_cdecl("sk_physics_body_apply_impulse")
public func sk_physics_body_apply_impulse(_ bodyHandle: UnsafeMutableRawPointer?, _ dx: Double, _ dy: Double) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.applyImpulse(CGVector(dx: dx, dy: dy))
}

@_cdecl("sk_physics_body_apply_impulse_at_point")
public func sk_physics_body_apply_impulse_at_point(
    _ bodyHandle: UnsafeMutableRawPointer?,
    _ impulseDx: Double, _ impulseDy: Double,
    _ pointX: Double, _ pointY: Double
) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.applyImpulse(CGVector(dx: impulseDx, dy: impulseDy), at: CGPoint(x: pointX, y: pointY))
}

@_cdecl("sk_physics_body_apply_angular_impulse")
public func sk_physics_body_apply_angular_impulse(_ bodyHandle: UnsafeMutableRawPointer?, _ impulse: Double) {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return }
    body.applyAngularImpulse(CGFloat(impulse))
}

@_cdecl("sk_physics_body_all_contacted_bodies_count")
public func sk_physics_body_all_contacted_bodies_count(_ bodyHandle: UnsafeMutableRawPointer?) -> Int {
    guard let body: SKPhysicsBody = skBorrow(bodyHandle) else { return 0 }
    return body.allContactedBodies().count
}

File: src/physics_body.rs
use core::ffi::c_void;

use apple_cf::cg::{CGPoint, CGRect, CGSize, CGVector};

use crate::ffi;
use crate::private::handle_type;
use crate::texture::Texture;

handle_type!(PhysicsBody);

impl PhysicsBody {
    /// Wraps `SKPhysicsBody`.
    #[must_use]
    pub fn circle(radius: f64) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_physics_body_circle(radius)) }
    }

    /// Wraps `SKPhysicsBody`.
    #[must_use]
    pub fn circle_at(radius: f64, center: CGPoint) -> Option<Self> {
        unsafe {
            Self::from_raw(ffi::sk_physics_body_circle_center(
                radius, center.x, center.y,
            ))
        }
    }

    /// Wraps `SKPhysicsBody`.
    #[must_use]
    pub fn rect(width: f64, height: f64) -> Option<Self> {
        unsafe { Self::from_raw(ffi::sk_physics_body_rect(width, height)) }
    }

    /// Wraps `SKPhysicsBody`.
    #[must_use]
    pub fn rect_at(size: CGSize, center: CGPoint) -> Option<Self> {
        unsafe {
            Self::from_raw(ffi::sk_physics_body_rect_center(
                size.width,
                size.height,
                center.x,
                center.y,
            ))
        }
    }

    /// Wraps `SKPhysicsBody`.
    #[must_use]
    pub fn edge_loop_rect(rect: CGRect) -> Option<Self> {
        unsafe {
            Self::from_raw(ffi::sk_physics_body_edge_loop_rect(
                rect.origin.x,
                rect.origin.y,
                rect.size.width,
                rect.size.height,
            ))
        }
    }

    /// Wraps `SKPhysicsBody`.
    #[must_use]
    pub fn from_texture(texture: &Texture, width: f64, height: f64) -> Option<Self> {
        unsafe {
            Self::from_raw(ffi::sk_physics_body_texture(
                texture.as_ptr(),
                width,
                height,
            ))
        }
    }

    /// Wraps `SKPhysicsBody`.
    #[must_use]
    pub fn compound(bodies: &[&Self]) -> Option<Self> {
        let mut raw: Vec<*mut c_void> = bodies.iter().map(|body| body.as_ptr()).collect();
        let raw_ptr = if raw.is_empty() {
            core::ptr::null_mut()
        } else {
            raw.as_mut_ptr().cast()
        };
        unsafe { Self::from_raw(ffi::sk_physics_body_compound(raw_ptr, raw.len())) }
    }

    /// Returns a property exposed by `SKPhysicsBody`.
    #[must_use]
    pub fn is_dynamic(&self) -> bool {
        unsafe { ffi::sk_physics_body_get_dynamic(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsBody`.
    pub fn set_dynamic(&self, dynamic: bool) {
        unsafe { ffi::sk_physics_body_set_dynamic(self.ptr, dynamic) };
    }

    /// Returns a property exposed by `SKPhysicsBody`.
    #[must_use]
    pub fn allows_rotation(&self) -> bool {
        unsafe { ffi::sk_physics_body_get_allows_rotation(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsBody`.
    pub fn set_allows_rotation(&self, allows: bool) {
        unsafe { ffi::sk_physics_body_set_allows_rotation(self.ptr, allows) };
    }

    /// Returns a property exposed by `SKPhysicsBody`.
    #[must_use]
    pub fn uses_precise_collision_detection(&self) -> bool {
        unsafe { ffi::sk_physics_body_get_precise_collision(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsBody`.
    pub fn set_uses_precise_collision_detection(&self, precise: bool) {
        unsafe { ffi::sk_physics_body_set_precise_collision(self.ptr, precise) };
    }

    /// Returns a property exposed by `SKPhysicsBody`.
    #[must_use]
    pub fn is_pinned(&self) -> bool {
        unsafe { ffi::sk_physics_body_get_pinned(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsBody`.
    pub fn set_pinned(&self, pinned: bool) {
        unsafe { ffi::sk_physics_body_set_pinned(self.ptr, pinned) };
    }

    /// Returns a property exposed by `SKPhysicsBody`.
    #[must_use]
    pub fn is_resting(&self) -> bool {
        unsafe { ffi::sk_physics_body_get_resting(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsBody`.
    pub fn set_resting(&self, resting: bool) {
        unsafe { ffi::sk_physics_body_set_resting(self.ptr, resting) };
    }

    /// Returns a property exposed by `SKPhysicsBody`.
    #[must_use]
    pub fn friction(&self) -> f64 {
        unsafe { ffi::sk_physics_body_get_friction(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsBody`.
    pub fn set_friction(&self, friction: f64) {
        unsafe { ffi::sk_physics_body_set_friction(self.ptr, friction) };
    }

    /// Returns a property exposed by `SKPhysicsBody`.
    #[must_use]
    pub fn charge(&self) -> f64 {
        unsafe { ffi::sk_physics_body_get_charge(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsBody`.
    pub fn set_charge(&self, charge: f64) {
        unsafe { ffi::sk_physics_body_set_charge(self.ptr, charge) };
    }

    /// Returns a property exposed by `SKPhysicsBody`.
    #[must_use]
    pub fn restitution(&self) -> f64 {
        unsafe { ffi::sk_physics_body_get_restitution(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsBody`.
    pub fn set_restitution(&self, restitution: f64) {
        unsafe { ffi::sk_physics_body_set_restitution(self.ptr, restitution) };
    }

    /// Returns a property exposed by `SKPhysicsBody`.
    #[must_use]
    pub fn linear_damping(&self) -> f64 {
        unsafe { ffi::sk_physics_body_get_linear_damping(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsBody`.
    pub fn set_linear_damping(&self, damping: f64) {
        unsafe { ffi::sk_physics_body_set_linear_damping(self.ptr, damping) };
    }

    /// Returns a property exposed by `SKPhysicsBody`.
    #[must_use]
    pub fn angular_damping(&self) -> f64 {
        unsafe { ffi::sk_physics_body_get_angular_damping(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsBody`.
    pub fn set_angular_damping(&self, damping: f64) {
        unsafe { ffi::sk_physics_body_set_angular_damping(self.ptr, damping) };
    }

    /// Returns a property exposed by `SKPhysicsBody`.
    #[must_use]
    pub fn density(&self) -> f64 {
        unsafe { ffi::sk_physics_body_get_density(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsBody`.
    pub fn set_density(&self, density: f64) {
        unsafe { ffi::sk_physics_body_set_density(self.ptr, density) };
    }

    /// Returns a property exposed by `SKPhysicsBody`.
    #[must_use]
    pub fn mass(&self) -> f64 {
        unsafe { ffi::sk_physics_body_get_mass(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsBody`.
    pub fn set_mass(&self, mass: f64) {
        unsafe { ffi::sk_physics_body_set_mass(self.ptr, mass) };
    }

    /// Returns a property exposed by `SKPhysicsBody`.
    #[must_use]
    pub fn area(&self) -> f64 {
        unsafe { ffi::sk_physics_body_get_area(self.ptr) }
    }

    /// Returns a property exposed by `SKPhysicsBody`.
    #[must_use]
    pub fn is_affected_by_gravity(&self) -> bool {
        unsafe { ffi::sk_physics_body_get_affected_by_gravity(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsBody`.
    pub fn set_affected_by_gravity(&self, affected: bool) {
        unsafe { ffi::sk_physics_body_set_affected_by_gravity(self.ptr, affected) };
    }

    /// Wraps `SKPhysicsBody`.
    #[must_use]
    pub fn field_bitmask(&self) -> u32 {
        unsafe { ffi::sk_physics_body_get_field_bitmask(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsBody`.
    pub fn set_field_bitmask(&self, mask: u32) {
        unsafe { ffi::sk_physics_body_set_field_bitmask(self.ptr, mask) };
    }

    /// Wraps `SKPhysicsBody`.
    #[must_use]
    pub fn category_bitmask(&self) -> u32 {
        unsafe { ffi::sk_physics_body_get_category_bitmask(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsBody`.
    pub fn set_category_bitmask(&self, mask: u32) {
        unsafe { ffi::sk_physics_body_set_category_bitmask(self.ptr, mask) };
    }

    /// Wraps `SKPhysicsBody`.
    #[must_use]
    pub fn contact_test_bitmask(&self) -> u32 {
        unsafe { ffi::sk_physics_body_get_contact_test_bitmask(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsBody`.
    pub fn set_contact_test_bitmask(&self, mask: u32) {
        unsafe { ffi::sk_physics_body_set_contact_test_bitmask(self.ptr, mask) };
    }

    /// Wraps `SKPhysicsBody`.
    #[must_use]
    pub fn collision_bitmask(&self) -> u32 {
        unsafe { ffi::sk_physics_body_get_collision_bitmask(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsBody`.
    pub fn set_collision_bitmask(&self, mask: u32) {
        unsafe { ffi::sk_physics_body_set_collision_bitmask(self.ptr, mask) };
    }

    /// Returns a property exposed by `SKPhysicsBody`.
    #[must_use]
    pub fn velocity(&self) -> CGVector {
        CGVector::new(
            unsafe { ffi::sk_physics_body_get_velocity_dx(self.ptr) },
            unsafe { ffi::sk_physics_body_get_velocity_dy(self.ptr) },
        )
    }

    /// Sets a property exposed by `SKPhysicsBody`.
    pub fn set_velocity(&self, velocity: CGVector) {
        unsafe { ffi::sk_physics_body_set_velocity(self.ptr, velocity.dx, velocity.dy) };
    }

    /// Returns a property exposed by `SKPhysicsBody`.
    #[must_use]
    pub fn angular_velocity(&self) -> f64 {
        unsafe { ffi::sk_physics_body_get_angular_velocity(self.ptr) }
    }

    /// Sets a property exposed by `SKPhysicsBody`.
    pub fn set_angular_velocity(&self, velocity: f64) {
        unsafe { ffi::sk_physics_body_set_angular_velocity(self.ptr, velocity) };
    }

    /// Wraps `SKPhysicsBody`.
    pub fn apply_force(&self, force: CGVector) {
        unsafe { ffi::sk_physics_body_apply_force(self.ptr, force.dx, force.dy) };
    }

    /// Wraps `SKPhysicsBody`.
    pub fn apply_force_at_point(&self, force: CGVector, point: CGPoint) {
        unsafe {
            ffi::sk_physics_body_apply_force_at_point(
                self.ptr, force.dx, force.dy, point.x, point.y,
            );
        };
    }

    /// Wraps `SKPhysicsBody`.
    pub fn apply_torque(&self, torque: f64) {
        unsafe { ffi::sk_physics_body_apply_torque(self.ptr, torque) };
    }

    /// Wraps `SKPhysicsBody`.
    pub fn apply_impulse(&self, impulse: CGVector) {
        unsafe { ffi::sk_physics_body_apply_impulse(self.ptr, impulse.dx, impulse.dy) };
    }

    /// Wraps `SKPhysicsBody`.
    pub fn apply_impulse_at_point(&self, impulse: CGVector, point: CGPoint) {
        unsafe {
            ffi::sk_physics_body_apply_impulse_at_point(
                self.ptr, impulse.dx, impulse.dy, point.x, point.y,
            );
        };
    }

    /// Wraps `SKPhysicsBody`.
    pub fn apply_angular_impulse(&self, impulse: f64) {
        unsafe { ffi::sk_physics_body_apply_angular_impulse(self.ptr, impulse) };
    }

    /// Returns a property exposed by `SKPhysicsBody`.
    #[must_use]
    pub fn all_contacted_bodies_count(&self) -> usize {
        unsafe { ffi::sk_physics_body_all_contacted_bodies_count(self.ptr) }
    }
}
